zcash_client_sqlite/src/wallet/sapling.rs

```diff
@@ -132,14 +132,34 @@ pub(crate) fn get_spendable_sapling_notes(
     anchor_height: BlockHeight,
     exclude: &[ReceivedNoteId],
 ) -> Result<Vec<ReceivedSaplingNote<ReceivedNoteId>>, SqliteClientError> {
+    let mut stmt_unscanned_tip = conn.prepare_cached(
+        "SELECT 1 FROM v_sapling_shard_unscanned_ranges unscanned
+         WHERE unscanned.shard_index IN (SELECT MAX(shard_index) FROM sapling_tree_shards)
+         AND unscanned.subtree_start_height <= :anchor_height",
+    )?;
+    let mut unscanned =
+        stmt_unscanned_tip.query(named_params![":anchor_height": &u32::from(anchor_height),])?;
+    if unscanned.next()?.is_some() {
+        // if the tip shard has unscanned ranges below the anchor height, none of our notes can be
+        // spent
+        return Ok(vec![]);
+    }
+
     let mut stmt_select_notes = conn.prepare_cached(
         "SELECT id_note, diversifier, value, rcm, commitment_tree_position
          FROM sapling_received_notes
          INNER JOIN transactions ON transactions.id_tx = sapling_received_notes.tx
          WHERE account = :account
+         AND commitment_tree_position IS NOT NULL
          AND spent IS NULL
          AND transactions.block <= :anchor_height
-         AND id_note NOT IN rarray(:exclude)",
+         AND id_note NOT IN rarray(:exclude)
+         AND NOT EXISTS (
+            SELECT 1 FROM v_sapling_shard_unscanned_ranges unscanned
+            WHERE sapling_received_notes.commitment_tree_position >= unscanned.start_position
+            AND sapling_received_notes.commitment_tree_position < unscanned.end_position_exclusive
+            AND unscanned.block_range_start <= :anchor_height
```

Comment:
```suggestion
```
This should be redundant because `AND transactions.block <= :anchor_height` already excludes notes received after the anchor height.

If there are any unscanned ranges between the position of the note and the anchor position, then we won't be able to build a witness for that note. So we need some other check that is checking whether the range from commitment_tree_position to the anchor position overlaps any unscanned range.

Reply:
@daira this is incorrect; we may have complete unscanned subtrees but we can still create a witness because we have the note commitment tree roots. We only need the subtree containing the note and the subtree containing the anchor to be fully scanned.